mlflow/llama_index/__init__.py

```diff
@@ -348,32 +354,158 @@ def _load_pyfunc(path, model_config: Optional[Dict[str, Any]] = None):
 
 
 class _LlamaIndexModelWrapper:
-    def __init__(self, index, engine_type, model_config):
+    def _build_engine_method(self) -> Callable:
+        if self.engine_type == _QUERY_ENGINE_NAME:
+            return self.index.as_query_engine(**self.model_config).query
+        elif self.engine_type == _CHAT_ENGINE_NAME:
+            return self.index.as_chat_engine(**self.model_config).chat
+        elif self.engine_type == _RETRIEVER_ENGINE_NAME:
+            return self.index.as_retriever(**self.model_config).retrieve
+        else:
+            raise ValueError(
+                f"Unsupported engine type: {self.engine_type}. It must be one of "
+                "{_SUPPORTED_ENGINES}"
+            )
+
+    def __init__(self, index, engine_type: str, model_config: Optional[Dict[str, Any]] = None):
+        self.index = index
+        self.model_config = model_config or {}
+        self.engine_type = engine_type
+        self.predict_callable = self._build_engine_method()
+
+    def _format_predict_input_query_and_retriever(self, data):
```

Comment:
Is it possible to make this function only accept single data and outputs `QueryBundle` only? Since the prediction is made one by one, I think it's fine to do formatting one by one too, and I believe it makes this function simpler?

This is introspection from LangChain experience, having multiple possible data types (e.g. batch vs single) flying within the data flow makes it super hard to reason about the logic and test with high coverage.

Reply:
Really great point!